### ISSUE TITLE
memory leak?

### ISSUE THREAD

**Issue by notestaff** [refresh-bio/KMC#94]
at https://github.com/refresh-bio/KMC/blob/master/kmer_counter/kb_storer.cpp#L79
should delete the individual elements of files[] allocated at
https://github.com/refresh-bio/KMC/blob/master/kmer_counter/kb_storer.cpp#L188
?
@marekkokot

**Comment by marekkokot** [comment]
Thanks,
It seems you are right and this memory is not released. 
After a quick look I think it should be freed rather here: https://github.com/refresh-bio/KMC/blob/master/kmer_counter/kb_reader.h#L215

This is because created objects are pointed not only inside array but they are also saved in some queue for future use. But I need to think this deeper to be sure.

**Comment by notestaff** [comment]
@marekkokot   maybe you could use https://en.cppreference.com/w/cpp/memory/shared_ptr , then don't need to think about where things get freed.   see also https://stackoverflow.com/questions/13061979/shared-ptr-to-an-array-should-it-be-used

**Comment by marekkokot** [comment]
Yeah, smart pointers are for sure good choice, though I would prefer to use rather unique_ptr that would be kept in the mentioned queue (as far as I remember it may be treated as the owner of those objects) and in the array (which could be replaced by vector in fact) I would store raw pointers. 
The problem is that when I started to write kmc_tools I have used smart pointers in some (rather not very important) places. And after release, there were problems because people had older compilers. The other problem is that probably there are more spots in KMC code that could use smart pointers instead of the old way new, delete approach. The problem is it would require rethinking many things, that are often far from trivial (I would really like to not use shared_ptr when unique_ptr may be used, btw very interesting talk in this subject https://www.youtube.com/watch?v=JfmTagWcqoE). It would also not introduce any new functionalities, so from the user perspective, I think it is not that important. Although I really like smart pointers and other nice features that are introduced in C++11 and 14 I don't think I will be able to make this kind of changes, but of course, memory leaks must be solved.
